src/contrib/cluster/Akka.Cluster.Tools/Singleton/ClusterSingletonProxy.cs

```diff
@@ -221,31 +223,27 @@ private void IdentifySingleton()
                 message: TryToIdentifySingleton.Instance,
                 sender: Self);
         }
-
-        private void TrackChanges(Action block)
-        {
-            var before = _membersByAge.FirstOrDefault();
-            block();
-            var after = _membersByAge.FirstOrDefault();
-
-            // if the head has changed, I need to find the new singleton
-            if (!Equals(before, after)) IdentifySingleton();
-        }
-
+        
         private void Add(Member member)
         {
             if (MatchingRole(member))
-                TrackChanges(() =>
-                {
-                    _membersByAge = _membersByAge.Remove(member); //replace
-                    _membersByAge = _membersByAge.Add(member);
-                });
+            {
+                var (newState, oldestChanged) = _state.AddMember(member);
+                _state = newState;
+                if (oldestChanged)
```

Comment:
Check to see if the oldest has changed, if so retry identifying the singleton (consistent with the behavior we had before, just no `Action` delegates and the `TrackChanges` method to do it.)